// @strapi/email/dist/admin/chunks/index-BPAZGe7V.mjs
const PERMISSIONS = {
    // This permission regards the main component (App) and is used to tell
    // If the plugin link should be displayed in the menu
    // And also if the plugin is accessible. This use case is found when a user types the url of the
    // plugin directly in the browser
    settings: [
        {
            action: 'plugin::email.settings.read',
            subject: null
        }
    ]
};

const prefixPluginTranslations = (trad, pluginId)=>{
    return Object.keys(trad).reduce((acc, current)=>{
        acc[`${pluginId}.${current}`] = trad[current];
        return acc;
    }, {});
};

function __variableDynamicImportRuntime1__(path) {
  switch (path) {
    case './translations/ar.json': return import('./ar-JWJKrFM-.mjs');
    case './translations/cs.json': return import('./cs-BZCMeXgq.mjs');
    case './translations/de.json': return import('./de-CGlAhulB.mjs');
    case './translations/dk.json': return import('./dk-B5IuEReS.mjs');
    case './translations/en.json': return import('./en-Dxg2wKLN.mjs');
    case './translations/es.json': return import('./es-fdc21M_E.mjs');
    case './translations/fr.json': return import('./fr-DZjvYp6C.mjs');
    case './translations/id.json': return import('./id-C1eDxX9W.mjs');
    case './translations/it.json': return import('./it-DVlJ2EhA.mjs');
    case './translations/ja.json': return import('./ja-CVosbw2M.mjs');
    case './translations/ko.json': return import('./ko-B8SR--vZ.mjs');
    case './translations/ms.json': return import('./ms-DLngEN2_.mjs');
    case './translations/nl.json': return import('./nl-B6Q0VF48.mjs');
    case './translations/pl.json': return import('./pl-Bve39qfV.mjs');
    case './translations/pt-BR.json': return import('./pt-BR-DPqDE4wA.mjs');
    case './translations/pt.json': return import('./pt-DLqF-Sb4.mjs');
    case './translations/ru.json': return import('./ru-rmFYdvwh.mjs');
    case './translations/sk.json': return import('./sk-DDJmUHcx.mjs');
    case './translations/th.json': return import('./th-BqTCSpUc.mjs');
    case './translations/tr.json': return import('./tr-BmrcZCoj.mjs');
    case './translations/uk.json': return import('./uk-Dsq-0GyK.mjs');
    case './translations/vi.json': return import('./vi-CO9gCtmD.mjs');
    case './translations/zh-Hans.json': return import('./zh-Hans-ChWc00HT.mjs');
    case './translations/zh.json': return import('./zh-CZQ4DV5L.mjs');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }
const admin = {
    // TODO typing app in strapi/types as every plugin needs it
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    register (app) {
        // Create the email settings section
        app.createSettingSection({
            id: 'email',
            intlLabel: {
                id: 'email.SettingsNav.section-label',
                defaultMessage: 'Email Plugin'
            }
        }, [
            {
                intlLabel: {
                    id: 'email.Settings.email.plugin.title',
                    defaultMessage: 'Settings'
                },
                id: 'settings',
                to: 'email',
                Component: ()=>import('./Settings-PuQuVXHu.mjs').then((mod)=>({
                            default: mod.ProtectedSettingsPage
                        })),
                permissions: PERMISSIONS.settings
            }
        ]);
        app.registerPlugin({
            id: 'email',
            name: 'email'
        });
    },
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    bootstrap () {},
    async registerTrads ({ locales }) {
        const importedTrads = await Promise.all(locales.map((locale)=>{
            return __variableDynamicImportRuntime1__(`./translations/${locale}.json`).then(({ default: data })=>{
                return {
                    data: prefixPluginTranslations(data, 'email'),
                    locale
                };
            }).catch(()=>{
                return {
                    data: {},
                    locale
                };
            });
        }));
        return Promise.resolve(importedTrads);
    }
};

export { PERMISSIONS as P, admin as a };
