// @strapi/content-manager/dist/admin/chunks/EditViewPage-BAI2r49P.mjs
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import * as React from 'react';
import { useQueryParams, useNotification, Page, Form, Blocker, useRBAC } from '@strapi/admin/strapi-admin';
import { Main, Tabs, Grid } from '@strapi/design-system';
import { useIntl } from 'react-intl';
import { useLocation, useParams } from 'react-router-dom';
import { styled } from 'styled-components';
import { c as useDoc, f as useDocumentRBAC, h as useDocumentLayout, i as createYupSchema, j as Header, g as getTranslation, P as Panels, k as PERMISSIONS, l as DocumentRBAC, S as SINGLE_TYPES } from './index-CcJLBLNf.mjs';
import { u as useLazyComponents, F as FormLayout } from './Input-BMLRZBE3.mjs';
import '@strapi/icons';
import 'lodash/fp/mapValues';
import 'yup';
import 'fractional-indexing';
import 'lodash/fp/pipe';
import 'qs';
import 'date-fns';
import '@reduxjs/toolkit';
import 'prismjs';
import 'slate';
import 'slate-history';
import 'slate-react';
import 'prismjs/themes/prism-solarizedlight.css';
import 'prismjs/components/prism-asmatmel';
import 'prismjs/components/prism-bash';
import 'prismjs/components/prism-basic';
import 'prismjs/components/prism-c';
import 'prismjs/components/prism-clojure';
import 'prismjs/components/prism-cobol';
import 'prismjs/components/prism-cpp';
import 'prismjs/components/prism-csharp';
import 'prismjs/components/prism-dart';
import 'prismjs/components/prism-docker';
import 'prismjs/components/prism-elixir';
import 'prismjs/components/prism-erlang';
import 'prismjs/components/prism-fortran';
import 'prismjs/components/prism-fsharp';
import 'prismjs/components/prism-go';
import 'prismjs/components/prism-graphql';
import 'prismjs/components/prism-groovy';
import 'prismjs/components/prism-haskell';
import 'prismjs/components/prism-haxe';
import 'prismjs/components/prism-ini';
import 'prismjs/components/prism-java';
import 'prismjs/components/prism-javascript';
import 'prismjs/components/prism-jsx';
import 'prismjs/components/prism-json';
import 'prismjs/components/prism-julia';
import 'prismjs/components/prism-kotlin';
import 'prismjs/components/prism-latex';
import 'prismjs/components/prism-lua';
import 'prismjs/components/prism-markdown';
import 'prismjs/components/prism-matlab';
import 'prismjs/components/prism-makefile';
import 'prismjs/components/prism-objectivec';
import 'prismjs/components/prism-perl';
import 'prismjs/components/prism-php';
import 'prismjs/components/prism-powershell';
import 'prismjs/components/prism-python';
import 'prismjs/components/prism-r';
import 'prismjs/components/prism-ruby';
import 'prismjs/components/prism-rust';
import 'prismjs/components/prism-sas';
import 'prismjs/components/prism-scala';
import 'prismjs/components/prism-scheme';
import 'prismjs/components/prism-sql';
import 'prismjs/components/prism-stata';
import 'prismjs/components/prism-swift';
import 'prismjs/components/prism-typescript';
import 'prismjs/components/prism-tsx';
import 'prismjs/components/prism-vbnet';
import 'prismjs/components/prism-yaml';
import './usePrev-Bjw2dhmq.mjs';
import './useDragAndDrop-HYwNDExe.mjs';
import 'react-dnd';
import '@radix-ui/react-toolbar';
import 'react-dnd-html5-backend';
import './objects-wl73iEma.mjs';
import 'lodash/clone';
import 'lodash/toPath';
import './Relations-CiOfFNxW.mjs';
import 'react-window';
import './relations-D1R7vM_e.mjs';
import './ComponentIcon-BZcTc4rj.mjs';
import '@strapi/icons/symbols';
import 'codemirror5';
import 'sanitize-html';
import 'highlight.js';
import 'markdown-it';
import 'markdown-it-abbr';
import 'markdown-it-container';
import 'markdown-it-deflist';
import 'markdown-it-emoji';
import 'markdown-it-footnote';
import 'markdown-it-ins';
import 'markdown-it-mark';
import 'markdown-it-sub';
import 'markdown-it-sup';
import 'highlight.js/styles/solarized-dark.css';
import 'codemirror5/addon/display/placeholder';

const useOnce = (effect)=>React.useEffect(effect, emptyDeps);
const emptyDeps = [];

/* -------------------------------------------------------------------------------------------------
 * EditViewPage
 * -----------------------------------------------------------------------------------------------*/ const EditViewPage = ()=>{
    const location = useLocation();
    const [{ query: { status } }, setQuery] = useQueryParams({
        status: 'draft'
    });
    const { formatMessage } = useIntl();
    const { toggleNotification } = useNotification();
    const { document, meta, isLoading: isLoadingDocument, schema, components, collectionType, id, model, hasError, getTitle, getInitialFormValues } = useDoc();
    const hasDraftAndPublished = schema?.options?.draftAndPublish ?? false;
    useOnce(()=>{
        /**
     * We only ever want to fire the notification once otherwise
     * whenever the app re-renders it'll pop up regardless of
     * what we do because the state comes from react-router-dom
     */ if (location?.state && 'error' in location.state) {
            toggleNotification({
                type: 'danger',
                message: location.state.error,
                timeout: 5000
            });
        }
    });
    const isLoadingActionsRBAC = useDocumentRBAC('EditViewPage', (state)=>state.isLoading);
    const isSingleType = collectionType === SINGLE_TYPES;
    /**
   * single-types don't current have an id, but because they're a singleton
   * we can simply use the update operation to continuously update the same
   * document with varying params.
   */ const isCreatingDocument = !id && !isSingleType;
    const { isLoading: isLoadingLayout, edit: { layout, settings: { mainField } } } = useDocumentLayout(model);
    const { isLazyLoading } = useLazyComponents([]);
    const isLoading = isLoadingActionsRBAC || isLoadingDocument || isLoadingLayout || isLazyLoading;
    const initialValues = getInitialFormValues(isCreatingDocument);
    if (isLoading && !document?.documentId) {
        return /*#__PURE__*/ jsx(Page.Loading, {});
    }
    if (!initialValues || hasError) {
        return /*#__PURE__*/ jsx(Page.Error, {});
    }
    const handleTabChange = (status)=>{
        if (status === 'published' || status === 'draft') {
            setQuery({
                status
            }, 'push', true);
        }
    };
    const validateSync = (values, options)=>{
        const yupSchema = createYupSchema(schema?.attributes, components, {
            status,
            ...options
        });
        return yupSchema.validateSync(values, {
            abortEarly: false
        });
    };
    return /*#__PURE__*/ jsxs(Main, {
        paddingLeft: 10,
        paddingRight: 10,
        children: [
            /*#__PURE__*/ jsx(Page.Title, {
                children: getTitle(mainField)
            }),
            /*#__PURE__*/ jsx(Form, {
                disabled: hasDraftAndPublished && status === 'published',
                initialValues: initialValues,
                method: isCreatingDocument ? 'POST' : 'PUT',
                validate: (values, options)=>{
                    const yupSchema = createYupSchema(schema?.attributes, components, {
                        status,
                        ...options
                    });
                    return yupSchema.validate(values, {
                        abortEarly: false
                    });
                },
                initialErrors: location?.state?.forceValidation ? validateSync(initialValues, {}) : {},
                children: ({ resetForm })=>/*#__PURE__*/ jsxs(Fragment, {
                        children: [
                            /*#__PURE__*/ jsx(Header, {
                                isCreating: isCreatingDocument,
                                status: hasDraftAndPublished ? getDocumentStatus(document, meta) : undefined,
                                title: getTitle(mainField)
                            }),
                            /*#__PURE__*/ jsxs(Tabs.Root, {
                                variant: "simple",
                                value: status,
                                onValueChange: handleTabChange,
                                children: [
                                    /*#__PURE__*/ jsx(Tabs.List, {
                                        "aria-label": formatMessage({
                                            id: getTranslation('containers.edit.tabs.label'),
                                            defaultMessage: 'Document status'
                                        }),
                                        children: hasDraftAndPublished ? /*#__PURE__*/ jsxs(Fragment, {
                                            children: [
                                                /*#__PURE__*/ jsx(StatusTab, {
                                                    value: "draft",
                                                    children: formatMessage({
                                                        id: getTranslation('containers.edit.tabs.draft'),
                                                        defaultMessage: 'draft'
                                                    })
                                                }),
                                                /*#__PURE__*/ jsx(StatusTab, {
                                                    disabled: !meta || meta.availableStatus.length === 0,
                                                    value: "published",
                                                    children: formatMessage({
                                                        id: getTranslation('containers.edit.tabs.published'),
                                                        defaultMessage: 'published'
                                                    })
                                                })
                                            ]
                                        }) : null
                                    }),
                                    /*#__PURE__*/ jsxs(Grid.Root, {
                                        paddingTop: 8,
                                        gap: 4,
                                        children: [
                                            /*#__PURE__*/ jsxs(Grid.Item, {
                                                col: 9,
                                                s: 12,
                                                direction: "column",
                                                alignItems: "stretch",
                                                children: [
                                                    /*#__PURE__*/ jsx(Tabs.Content, {
                                                        value: "draft",
                                                        children: /*#__PURE__*/ jsx(FormLayout, {
                                                            layout: layout
                                                        })
                                                    }),
                                                    /*#__PURE__*/ jsx(Tabs.Content, {
                                                        value: "published",
                                                        children: /*#__PURE__*/ jsx(FormLayout, {
                                                            layout: layout
                                                        })
                                                    })
                                                ]
                                            }),
                                            /*#__PURE__*/ jsx(Grid.Item, {
                                                col: 3,
                                                s: 12,
                                                direction: "column",
                                                alignItems: "stretch",
                                                children: /*#__PURE__*/ jsx(Panels, {})
                                            })
                                        ]
                                    })
                                ]
                            }),
                            /*#__PURE__*/ jsx(Blocker, {
                                // We reset the form to the published version to avoid errors like – https://strapi-inc.atlassian.net/browse/CONTENT-2284
                                onProceed: resetForm
                            })
                        ]
                    })
            })
        ]
    });
};
const StatusTab = styled(Tabs.Trigger)`
  text-transform: uppercase;
`;
/**
 * @internal
 * @description Returns the status of the document where its latest state takes priority,
 * this typically will be "published" unless a user has edited their draft in which we should
 * display "modified".
 */ const getDocumentStatus = (document, meta)=>{
    const docStatus = document?.status;
    const statuses = meta?.availableStatus ?? [];
    /**
   * Creating an entry
   */ if (!docStatus) {
        return 'draft';
    }
    /**
   * We're viewing a draft, but the document could have a published version
   */ if (docStatus === 'draft' && statuses.find((doc)=>doc.publishedAt !== null)) {
        return 'published';
    }
    return docStatus;
};
/* -------------------------------------------------------------------------------------------------
 * ProtectedEditViewPage
 * -----------------------------------------------------------------------------------------------*/ const ProtectedEditViewPage = ()=>{
    const { slug = '' } = useParams();
    const { permissions = [], isLoading, error } = useRBAC(PERMISSIONS.map((action)=>({
            action,
            subject: slug
        })));
    if (isLoading) {
        return /*#__PURE__*/ jsx(Page.Loading, {});
    }
    if (error || !slug) {
        return /*#__PURE__*/ jsx(Page.Error, {});
    }
    return /*#__PURE__*/ jsx(Page.Protect, {
        permissions: permissions,
        children: ({ permissions })=>/*#__PURE__*/ jsx(DocumentRBAC, {
                permissions: permissions,
                children: /*#__PURE__*/ jsx(EditViewPage, {})
            })
    });
};

export { EditViewPage, ProtectedEditViewPage, getDocumentStatus };
