// @strapi/content-manager/dist/admin/chunks/EditViewPage-DpvonssD.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var reactIntl = require('react-intl');
var reactRouterDom = require('react-router-dom');
var styledComponents = require('styled-components');
var index = require('./index-SQ88CePz.js');
var Input = require('./Input-Bv-rqfYH.js');
require('@strapi/icons');
require('lodash/fp/mapValues');
require('yup');
require('fractional-indexing');
require('lodash/fp/pipe');
require('qs');
require('date-fns');
require('@reduxjs/toolkit');
require('prismjs');
require('slate');
require('slate-history');
require('slate-react');
require('prismjs/themes/prism-solarizedlight.css');
require('prismjs/components/prism-asmatmel');
require('prismjs/components/prism-bash');
require('prismjs/components/prism-basic');
require('prismjs/components/prism-c');
require('prismjs/components/prism-clojure');
require('prismjs/components/prism-cobol');
require('prismjs/components/prism-cpp');
require('prismjs/components/prism-csharp');
require('prismjs/components/prism-dart');
require('prismjs/components/prism-docker');
require('prismjs/components/prism-elixir');
require('prismjs/components/prism-erlang');
require('prismjs/components/prism-fortran');
require('prismjs/components/prism-fsharp');
require('prismjs/components/prism-go');
require('prismjs/components/prism-graphql');
require('prismjs/components/prism-groovy');
require('prismjs/components/prism-haskell');
require('prismjs/components/prism-haxe');
require('prismjs/components/prism-ini');
require('prismjs/components/prism-java');
require('prismjs/components/prism-javascript');
require('prismjs/components/prism-jsx');
require('prismjs/components/prism-json');
require('prismjs/components/prism-julia');
require('prismjs/components/prism-kotlin');
require('prismjs/components/prism-latex');
require('prismjs/components/prism-lua');
require('prismjs/components/prism-markdown');
require('prismjs/components/prism-matlab');
require('prismjs/components/prism-makefile');
require('prismjs/components/prism-objectivec');
require('prismjs/components/prism-perl');
require('prismjs/components/prism-php');
require('prismjs/components/prism-powershell');
require('prismjs/components/prism-python');
require('prismjs/components/prism-r');
require('prismjs/components/prism-ruby');
require('prismjs/components/prism-rust');
require('prismjs/components/prism-sas');
require('prismjs/components/prism-scala');
require('prismjs/components/prism-scheme');
require('prismjs/components/prism-sql');
require('prismjs/components/prism-stata');
require('prismjs/components/prism-swift');
require('prismjs/components/prism-typescript');
require('prismjs/components/prism-tsx');
require('prismjs/components/prism-vbnet');
require('prismjs/components/prism-yaml');
require('./usePrev-DIYl-IAL.js');
require('./useDragAndDrop-gcqEJMnO.js');
require('react-dnd');
require('@radix-ui/react-toolbar');
require('react-dnd-html5-backend');
require('./objects-C3EebVVe.js');
require('lodash/clone');
require('lodash/toPath');
require('./Relations-CJ0GWuqq.js');
require('react-window');
require('./relations-VlsO9KQZ.js');
require('./ComponentIcon-C-EjOUPA.js');
require('@strapi/icons/symbols');
require('codemirror5');
require('sanitize-html');
require('highlight.js');
require('markdown-it');
require('markdown-it-abbr');
require('markdown-it-container');
require('markdown-it-deflist');
require('markdown-it-emoji');
require('markdown-it-footnote');
require('markdown-it-ins');
require('markdown-it-mark');
require('markdown-it-sub');
require('markdown-it-sup');
require('highlight.js/styles/solarized-dark.css');
require('codemirror5/addon/display/placeholder');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const useOnce = (effect)=>React__namespace.useEffect(effect, emptyDeps);
const emptyDeps = [];

/* -------------------------------------------------------------------------------------------------
 * EditViewPage
 * -----------------------------------------------------------------------------------------------*/ const EditViewPage = ()=>{
    const location = reactRouterDom.useLocation();
    const [{ query: { status } }, setQuery] = strapiAdmin.useQueryParams({
        status: 'draft'
    });
    const { formatMessage } = reactIntl.useIntl();
    const { toggleNotification } = strapiAdmin.useNotification();
    const { document, meta, isLoading: isLoadingDocument, schema, components, collectionType, id, model, hasError, getTitle, getInitialFormValues } = index.useDoc();
    const hasDraftAndPublished = schema?.options?.draftAndPublish ?? false;
    useOnce(()=>{
        /**
     * We only ever want to fire the notification once otherwise
     * whenever the app re-renders it'll pop up regardless of
     * what we do because the state comes from react-router-dom
     */ if (location?.state && 'error' in location.state) {
            toggleNotification({
                type: 'danger',
                message: location.state.error,
                timeout: 5000
            });
        }
    });
    const isLoadingActionsRBAC = index.useDocumentRBAC('EditViewPage', (state)=>state.isLoading);
    const isSingleType = collectionType === index.SINGLE_TYPES;
    /**
   * single-types don't current have an id, but because they're a singleton
   * we can simply use the update operation to continuously update the same
   * document with varying params.
   */ const isCreatingDocument = !id && !isSingleType;
    const { isLoading: isLoadingLayout, edit: { layout, settings: { mainField } } } = index.useDocumentLayout(model);
    const { isLazyLoading } = Input.useLazyComponents([]);
    const isLoading = isLoadingActionsRBAC || isLoadingDocument || isLoadingLayout || isLazyLoading;
    const initialValues = getInitialFormValues(isCreatingDocument);
    if (isLoading && !document?.documentId) {
        return /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Page.Loading, {});
    }
    if (!initialValues || hasError) {
        return /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Page.Error, {});
    }
    const handleTabChange = (status)=>{
        if (status === 'published' || status === 'draft') {
            setQuery({
                status
            }, 'push', true);
        }
    };
    const validateSync = (values, options)=>{
        const yupSchema = index.createYupSchema(schema?.attributes, components, {
            status,
            ...options
        });
        return yupSchema.validateSync(values, {
            abortEarly: false
        });
    };
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Main, {
        paddingLeft: 10,
        paddingRight: 10,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Page.Title, {
                children: getTitle(mainField)
            }),
            /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Form, {
                disabled: hasDraftAndPublished && status === 'published',
                initialValues: initialValues,
                method: isCreatingDocument ? 'POST' : 'PUT',
                validate: (values, options)=>{
                    const yupSchema = index.createYupSchema(schema?.attributes, components, {
                        status,
                        ...options
                    });
                    return yupSchema.validate(values, {
                        abortEarly: false
                    });
                },
                initialErrors: location?.state?.forceValidation ? validateSync(initialValues, {}) : {},
                children: ({ resetForm })=>/*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsx(index.Header, {
                                isCreating: isCreatingDocument,
                                status: hasDraftAndPublished ? getDocumentStatus(document, meta) : undefined,
                                title: getTitle(mainField)
                            }),
                            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Tabs.Root, {
                                variant: "simple",
                                value: status,
                                onValueChange: handleTabChange,
                                children: [
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Tabs.List, {
                                        "aria-label": formatMessage({
                                            id: index.getTranslation('containers.edit.tabs.label'),
                                            defaultMessage: 'Document status'
                                        }),
                                        children: hasDraftAndPublished ? /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
                                            children: [
                                                /*#__PURE__*/ jsxRuntime.jsx(StatusTab, {
                                                    value: "draft",
                                                    children: formatMessage({
                                                        id: index.getTranslation('containers.edit.tabs.draft'),
                                                        defaultMessage: 'draft'
                                                    })
                                                }),
                                                /*#__PURE__*/ jsxRuntime.jsx(StatusTab, {
                                                    disabled: !meta || meta.availableStatus.length === 0,
                                                    value: "published",
                                                    children: formatMessage({
                                                        id: index.getTranslation('containers.edit.tabs.published'),
                                                        defaultMessage: 'published'
                                                    })
                                                })
                                            ]
                                        }) : null
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Grid.Root, {
                                        paddingTop: 8,
                                        gap: 4,
                                        children: [
                                            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Grid.Item, {
                                                col: 9,
                                                s: 12,
                                                direction: "column",
                                                alignItems: "stretch",
                                                children: [
                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Tabs.Content, {
                                                        value: "draft",
                                                        children: /*#__PURE__*/ jsxRuntime.jsx(Input.FormLayout, {
                                                            layout: layout
                                                        })
                                                    }),
                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Tabs.Content, {
                                                        value: "published",
                                                        children: /*#__PURE__*/ jsxRuntime.jsx(Input.FormLayout, {
                                                            layout: layout
                                                        })
                                                    })
                                                ]
                                            }),
                                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Grid.Item, {
                                                col: 3,
                                                s: 12,
                                                direction: "column",
                                                alignItems: "stretch",
                                                children: /*#__PURE__*/ jsxRuntime.jsx(index.Panels, {})
                                            })
                                        ]
                                    })
                                ]
                            }),
                            /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Blocker, {
                                // We reset the form to the published version to avoid errors like – https://strapi-inc.atlassian.net/browse/CONTENT-2284
                                onProceed: resetForm
                            })
                        ]
                    })
            })
        ]
    });
};
const StatusTab = styledComponents.styled(designSystem.Tabs.Trigger)`
  text-transform: uppercase;
`;
/**
 * @internal
 * @description Returns the status of the document where its latest state takes priority,
 * this typically will be "published" unless a user has edited their draft in which we should
 * display "modified".
 */ const getDocumentStatus = (document, meta)=>{
    const docStatus = document?.status;
    const statuses = meta?.availableStatus ?? [];
    /**
   * Creating an entry
   */ if (!docStatus) {
        return 'draft';
    }
    /**
   * We're viewing a draft, but the document could have a published version
   */ if (docStatus === 'draft' && statuses.find((doc)=>doc.publishedAt !== null)) {
        return 'published';
    }
    return docStatus;
};
/* -------------------------------------------------------------------------------------------------
 * ProtectedEditViewPage
 * -----------------------------------------------------------------------------------------------*/ const ProtectedEditViewPage = ()=>{
    const { slug = '' } = reactRouterDom.useParams();
    const { permissions = [], isLoading, error } = strapiAdmin.useRBAC(index.PERMISSIONS.map((action)=>({
            action,
            subject: slug
        })));
    if (isLoading) {
        return /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Page.Loading, {});
    }
    if (error || !slug) {
        return /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Page.Error, {});
    }
    return /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Page.Protect, {
        permissions: permissions,
        children: ({ permissions })=>/*#__PURE__*/ jsxRuntime.jsx(index.DocumentRBAC, {
                permissions: permissions,
                children: /*#__PURE__*/ jsxRuntime.jsx(EditViewPage, {})
            })
    });
};

exports.EditViewPage = EditViewPage;
exports.ProtectedEditViewPage = ProtectedEditViewPage;
exports.getDocumentStatus = getDocumentStatus;
