// @strapi/admin/dist/admin/index.mjs
export { B as BackButton, I as Blocker, C as ConfirmDialog, v as ContentBox, H as DescriptionComponentRenderer, s as Filters, F as Form, M as InputRenderer, b as Layouts, P as Pagination, S as SearchInput, T as Table, J as getYupValidationErrors, G as renderAdmin, e as translatedErrors, k as useAdminUsers, t as useClipboard, E as useElementOnScreen, n as useFetchClient, u as useField, D as useFocusInputField, w as useForm, z as useHistory, A as useInjectReducer, K as useTable } from './chunks/index-GvH7eseL.mjs';
export { N as NotificationsProvider, P as Page, z as createContext, e as useAPIErrorHandler, p as useAppInfo, u as useAuth, m as useGuidedTour, a as useNotification, n as useQueryParams, k as useRBAC, q as useStrapiApp, c as useTracking } from './chunks/Theme-C3oHrQcy.mjs';
export { F as FetchError, a as adminApi, l as fetchBaseQuery, k as getFetchClient, i as isBaseQueryError, c as isFetchError } from './chunks/admin-BBS7sDVu.mjs';
import 'react-dom/client';
import 'react/jsx-runtime';
import 'react';
import '@strapi/design-system';
import 'invariant';
import 'lodash/isFunction';
import 'lodash/merge';
import 'lodash/pick';
import 'react-router-dom';
import 'immer';
import 'react-redux';
import 'react-query';
import '@strapi/icons';
import 'react-intl';
import 'styled-components';
import '@strapi/icons/symbols';
import './chunks/useEnterprise-Dy8wMNV5.mjs';
import 'lodash/camelCase';
import 'yup';
import 'fractional-indexing';
import 'lodash/isEqual';
import 'lodash/omit';
import 'formik';
import '@reduxjs/toolkit';
import 'lodash/throttle';
import 'qs';
import '@radix-ui/react-context';
import 'lodash/clone';
import 'lodash/toPath';
import 'axios';
import 'lodash/get';
import 'lodash/set';
import 'use-context-selector';
import 'lodash/defaultsDeep';
import 'lodash/trimEnd';
import '@reduxjs/toolkit/query/react';
import 'lodash/fp/pipe';
