// @strapi/content-manager/dist/shared/index.js
'use strict';

var collectionTypes = /*#__PURE__*/Object.freeze({
  __proto__: null
});

var components = /*#__PURE__*/Object.freeze({
  __proto__: null
});

var contentTypes = /*#__PURE__*/Object.freeze({
  __proto__: null
});

var init = /*#__PURE__*/Object.freeze({
  __proto__: null
});

var relations = /*#__PURE__*/Object.freeze({
  __proto__: null
});

var singleTypes = /*#__PURE__*/Object.freeze({
  __proto__: null
});

var uid = /*#__PURE__*/Object.freeze({
  __proto__: null
});

var reviewWorkflows = /*#__PURE__*/Object.freeze({
  __proto__: null
});

var historyVersions = /*#__PURE__*/Object.freeze({
  __proto__: null
});

var preview = /*#__PURE__*/Object.freeze({
  __proto__: null
});

var index = /*#__PURE__*/Object.freeze({
  __proto__: null,
  CollectionTypes: collectionTypes,
  Components: components,
  ContentTypes: contentTypes,
  HistoryVersions: historyVersions,
  Init: init,
  Preview: preview,
  Relations: relations,
  ReviewWorkflows: reviewWorkflows,
  SingleTypes: singleTypes,
  UID: uid
});

exports.Contracts = index;
