// @strapi/logger/dist/index.mjs
import * as winston from 'winston';
import { config, format, transports } from 'winston';
export { winston };

const LEVELS = config.npm.levels;
const LEVEL_LABEL = 'silly';
LEVELS[LEVEL_LABEL];

const logErrors = format((info)=>{
    if (info instanceof Error) {
        return {
            ...info,
            message: `${info.message}${info.stack ? `\n${info.stack}` : ''}`
        };
    }
    return info;
});

const defaultTimestampFormat = 'YYYY-MM-DD HH:mm:ss.SSS';
/**
 * Create a pretty print formatter for a winston logger
 * @param options
 */ var prettyPrint = ((options = {})=>{
    const { timestamps = true, colors = true } = options;
    const handlers = [];
    if (timestamps) {
        handlers.push(format.timestamp({
            format: timestamps === true ? defaultTimestampFormat : timestamps
        }));
    }
    if (colors) {
        handlers.push(format.colorize());
    }
    handlers.push(logErrors());
    handlers.push(format.printf(({ level, message, timestamp })=>{
        return `${timestamps ? `[${timestamp}] ` : ''}${level}: ${message}`;
    }));
    return format.combine(...handlers);
});

var levelFilter = ((...levels)=>{
    return format((info)=>levels.some((level)=>info.level.includes(level)) ? info : false)();
});

/**
 * This will remove the chalk color codes from the message provided.
 * It's used to log plain text in the log file
 */ var excludeColors = format.printf(({ message })=>{
    if (typeof message !== 'string') {
        return message;
    }
    return message.replace(// eslint-disable-next-line no-control-regex
    /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g, '');
});

/**
 * This will remove the chalk color codes from the message provided.
 * It's used to log plain text in the log file
 */ var detailedLog = format.printf(({ message, level, timestamp })=>{
    if (typeof message !== 'string') {
        return message;
    }
    const newMessage = `[${timestamp}] ${level}: ${message}`;
    return newMessage.replace(// eslint-disable-next-line no-control-regex
    /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g, '');
});

var index$1 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  detailedLogs: detailedLog,
  excludeColors: excludeColors,
  levelFilter: levelFilter,
  prettyPrint: prettyPrint
});

var defaultConfiguration = (()=>{
    return {
        level: LEVEL_LABEL,
        levels: LEVELS,
        format: prettyPrint(),
        transports: [
            new transports.Console()
        ]
    };
});

var outputFileConfiguration = ((filename, fileTransportOptions = {})=>{
    return {
        level: LEVEL_LABEL,
        levels: LEVELS,
        format: prettyPrint(),
        transports: [
            new transports.Console(),
            new transports.File({
                level: 'error',
                filename,
                format: excludeColors,
                ...fileTransportOptions
            })
        ]
    };
});

var index = /*#__PURE__*/Object.freeze({
  __proto__: null,
  createDefaultConfiguration: defaultConfiguration,
  createOutputFileConfiguration: outputFileConfiguration
});

const createLogger = (userConfiguration = {})=>{
    const configuration = defaultConfiguration();
    Object.assign(configuration, userConfiguration);
    return winston.createLogger(configuration);
};

export { index as configs, createLogger, index$1 as formats };
