// @strapi/admin/dist/admin/chunks/AdminSeatInfo-CQ4w8a3g.mjs
import { jsxs, jsx } from 'react/jsx-runtime';
import { Grid, Typography, Flex, Tooltip, Link } from '@strapi/design-system';
import { WarningCircle, ExternalLink } from '@strapi/icons';
import { useIntl } from 'react-intl';
import { useSelector } from 'react-redux';
import { k as useRBAC } from './Theme-C3oHrQcy.mjs';
import { s as selectAdminPermissions } from './selectors-Dej_6h4x.mjs';
import { u as useLicenseLimits } from './useLicenseLimits-D4XBewxi.mjs';
import 'react';
import '@radix-ui/react-context';
import '@strapi/icons/symbols';
import 'react-router-dom';
import 'qs';
import './admin-BBS7sDVu.mjs';
import 'lodash/trimEnd';
import '@reduxjs/toolkit/query/react';
import 'lodash/fp/pipe';
import 'lodash/clone';
import 'lodash/toPath';
import 'lodash/isEqual';
import 'axios';
import 'immer';
import 'lodash/get';
import 'lodash/set';
import 'use-context-selector';
import 'lodash/defaultsDeep';
import 'styled-components';
import '@reduxjs/toolkit';

const BILLING_STRAPI_CLOUD_URL = 'https://cloud.strapi.io/profile/billing';
const BILLING_SELF_HOSTED_URL = 'https://strapi.io/billing/request-seats';
const AdminSeatInfoEE = ()=>{
    const { formatMessage } = useIntl();
    const { settings } = useSelector(selectAdminPermissions);
    const { isLoading: isRBACLoading, allowedActions: { canRead, canCreate, canUpdate, canDelete } } = useRBAC(settings?.users ?? {});
    const { license, isError, isLoading: isLicenseLoading } = useLicenseLimits({
        // TODO: this creates a waterfall which we should avoid to render earlier, but for that
        // we will have to move away from data-fetching hooks to query functions.
        // Short-term we could at least implement a loader, for the user to have visual feedback
        // in case the requests take a while
        enabled: !isRBACLoading && canRead && canCreate && canUpdate && canDelete
    });
    const isLoading = isRBACLoading || isLicenseLoading;
    if (isError || isLoading || !license) {
        return null;
    }
    const { licenseLimitStatus, enforcementUserCount, permittedSeats, isHostedOnStrapiCloud } = license;
    if (!permittedSeats) {
        return null;
    }
    return /*#__PURE__*/ jsxs(Grid.Item, {
        col: 6,
        s: 12,
        direction: "column",
        alignItems: "stretch",
        children: [
            /*#__PURE__*/ jsx(Typography, {
                variant: "sigma",
                textColor: "neutral600",
                children: formatMessage({
                    id: 'Settings.application.admin-seats',
                    defaultMessage: 'Admin seats'
                })
            }),
            /*#__PURE__*/ jsxs(Flex, {
                gap: 2,
                children: [
                    /*#__PURE__*/ jsx(Flex, {
                        children: /*#__PURE__*/ jsx(Typography, {
                            tag: "p",
                            children: formatMessage({
                                id: 'Settings.application.ee.admin-seats.count',
                                defaultMessage: '<text>{enforcementUserCount}</text>/{permittedSeats}'
                            }, {
                                permittedSeats,
                                enforcementUserCount,
                                text: (chunks)=>/*#__PURE__*/ jsx(Typography, {
                                        fontWeight: "semiBold",
                                        textColor: enforcementUserCount > permittedSeats ? 'danger500' : undefined,
                                        children: chunks
                                    })
                            })
                        })
                    }),
                    licenseLimitStatus === 'OVER_LIMIT' && /*#__PURE__*/ jsx(Tooltip, {
                        label: formatMessage({
                            id: 'Settings.application.ee.admin-seats.at-limit-tooltip',
                            defaultMessage: 'At limit: add seats to invite more users'
                        }),
                        children: /*#__PURE__*/ jsx(WarningCircle, {
                            width: "1.4rem",
                            height: "1.4rem",
                            fill: "danger500"
                        })
                    })
                ]
            }),
            /*#__PURE__*/ jsx(Link, {
                href: isHostedOnStrapiCloud ? BILLING_STRAPI_CLOUD_URL : BILLING_SELF_HOSTED_URL,
                isExternal: true,
                endIcon: /*#__PURE__*/ jsx(ExternalLink, {}),
                children: formatMessage({
                    id: 'Settings.application.ee.admin-seats.add-seats',
                    defaultMessage: '{isHostedOnStrapiCloud, select, true {Add seats} other {Contact sales}}'
                }, {
                    isHostedOnStrapiCloud
                })
            })
        ]
    });
};

export { AdminSeatInfoEE };
