// @strapi/admin/dist/admin/chunks/AuthenticatedLayout-B67MSNKW.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var packageInfo = require('@strapi/admin/package.json');
var designSystem = require('@strapi/design-system');
var reactDnd = require('react-dnd');
var reactDndHtml5Backend = require('react-dnd-html5-backend');
var reactIntl = require('react-intl');
var reactRouterDom = require('react-router-dom');
var lt = require('semver/functions/lt');
var valid = require('semver/functions/valid');
var icons = require('@strapi/icons');
var get = require('lodash/get');
var styledComponents = require('styled-components');
var Theme = require('./Theme-DDlJz91O.js');
var Ornaments = require('./Ornaments-CvdERnok.js');
var users = require('./users-D-oKxjK9.js');
var index = require('./index-C8yOprVG.js');
var immer = require('immer');
var PrivateRoute = require('./PrivateRoute-DvGSyomB.js');
var cloneDeep = require('lodash/cloneDeep');
var useOnce = require('./useOnce-CCEXokcy.js');
var admin = require('./admin-udBiOv2o.js');
require('@radix-ui/react-context');
require('@strapi/icons/symbols');
require('qs');
require('lodash/clone');
require('lodash/toPath');
require('lodash/isEqual');
require('axios');
require('lodash/set');
require('use-context-selector');
require('lodash/defaultsDeep');
require('react-redux');
require('@reduxjs/toolkit');
require('react-dom/client');
require('invariant');
require('lodash/isFunction');
require('lodash/merge');
require('lodash/pick');
require('react-query');
require('./useEnterprise-IHd3AKd2.js');
require('lodash/camelCase');
require('yup');
require('fractional-indexing');
require('lodash/omit');
require('formik');
require('lodash/throttle');
require('lodash/trimEnd');
require('@reduxjs/toolkit/query/react');
require('lodash/fp/pipe');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

/* -------------------------------------------------------------------------------------------------
 * GuidedTourModal
 * -----------------------------------------------------------------------------------------------*/ const GuidedTourModal = ()=>{
    const guidedTour = Theme.useGuidedTour('GuidedTourModal', (state)=>state);
    const { currentStep, guidedTourState, setCurrentStep, setStepState, isGuidedTourVisible, setSkipped } = guidedTour;
    const { formatMessage } = reactIntl.useIntl();
    const { trackUsage } = Theme.useTracking();
    if (!currentStep || !isGuidedTourVisible) {
        return null;
    }
    const stepData = get(Ornaments.LAYOUT_DATA, currentStep);
    const sectionKeys = Object.keys(guidedTourState);
    const [sectionName, stepName] = currentStep.split('.');
    const sectionIndex = sectionKeys.indexOf(sectionName);
    const stepIndex = Object.keys(guidedTourState[sectionName]).indexOf(stepName);
    const hasSectionAfter = sectionIndex < sectionKeys.length - 1;
    const hasStepAfter = stepIndex < Object.keys(guidedTourState[sectionName]).length - 1;
    const handleCtaClick = ()=>{
        setStepState(currentStep, true);
        if (stepData) {
            trackUsage(stepData.trackingEvent);
        }
        setCurrentStep(null);
    };
    const handleSkip = ()=>{
        setSkipped(true);
        setCurrentStep(null);
        trackUsage('didSkipGuidedtour');
    };
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Portal, {
        children: /*#__PURE__*/ jsxRuntime.jsx(ModalWrapper, {
            onClick: handleCtaClick,
            padding: 8,
            justifyContent: "center",
            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.FocusTrap, {
                onEscape: handleCtaClick,
                children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                    direction: "column",
                    alignItems: "stretch",
                    background: "neutral0",
                    width: `66rem`,
                    shadow: "popupShadow",
                    hasRadius: true,
                    padding: 4,
                    gap: 8,
                    role: "dialog",
                    "aria-modal": true,
                    onClick: (e)=>e.stopPropagation(),
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                            justifyContent: "flex-end",
                            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                                onClick: handleCtaClick,
                                withTooltip: false,
                                label: formatMessage({
                                    id: 'app.utils.close-label',
                                    defaultMessage: 'Close'
                                }),
                                children: /*#__PURE__*/ jsxRuntime.jsx(icons.Cross, {})
                            })
                        }),
                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                            paddingLeft: 7,
                            paddingRight: 7,
                            paddingBottom: !hasStepAfter && !hasSectionAfter ? 8 : 0,
                            children: /*#__PURE__*/ jsxRuntime.jsx(GuidedTourStepper, {
                                title: stepData && 'title' in stepData ? stepData.title : undefined,
                                cta: stepData && 'cta' in stepData ? stepData.cta : undefined,
                                onCtaClick: handleCtaClick,
                                sectionIndex: sectionIndex,
                                stepIndex: stepIndex,
                                hasSectionAfter: hasSectionAfter,
                                children: stepData && 'content' in stepData && /*#__PURE__*/ jsxRuntime.jsx(GuidedTourContent, {
                                    ...stepData.content
                                })
                            })
                        }),
                        !(!hasStepAfter && !hasSectionAfter) && /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                            justifyContent: "flex-end",
                            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                                variant: "tertiary",
                                onClick: handleSkip,
                                children: formatMessage({
                                    id: 'app.components.GuidedTour.skip',
                                    defaultMessage: 'Skip the tour'
                                })
                            })
                        })
                    ]
                })
            })
        })
    });
};
const ModalWrapper = styledComponents.styled(designSystem.Flex)`
  position: fixed;
  z-index: 4;
  inset: 0;
  /* this is theme.colors.neutral800 with opacity */
  background: ${({ theme })=>`${theme.colors.neutral800}1F`};
`;
const GuidedTourStepper = ({ title, children, cta, onCtaClick, sectionIndex, stepIndex, hasSectionAfter })=>{
    const { formatMessage } = reactIntl.useIntl();
    const hasSectionBefore = sectionIndex > 0;
    const hasStepsBefore = stepIndex > 0;
    const nextSectionIndex = sectionIndex + 1;
    return /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
        children: [
            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                alignItems: "stretch",
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                        marginRight: 8,
                        justifyContent: "center",
                        minWidth: `3rem`,
                        children: hasSectionBefore && /*#__PURE__*/ jsxRuntime.jsx(Ornaments.VerticalDivider, {
                            state: Ornaments.STATES.IS_DONE,
                            minHeight: `2.4rem`
                        })
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                        variant: "sigma",
                        textColor: "primary600",
                        children: formatMessage({
                            id: 'app.components.GuidedTour.title',
                            defaultMessage: '3 steps to get started'
                        })
                    })
                ]
            }),
            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                        marginRight: 8,
                        minWidth: `3rem`,
                        children: /*#__PURE__*/ jsxRuntime.jsx(Ornaments.Number, {
                            state: hasStepsBefore ? Ornaments.STATES.IS_DONE : Ornaments.STATES.IS_ACTIVE,
                            paddingTop: 3,
                            paddingBottom: 3,
                            children: sectionIndex + 1
                        })
                    }),
                    title && /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                        variant: "alpha",
                        fontWeight: "bold",
                        textColor: "neutral800",
                        tag: "h3",
                        id: "title",
                        children: formatMessage(title)
                    })
                ]
            }),
            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                alignItems: "stretch",
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                        marginRight: 8,
                        direction: "column",
                        justifyContent: "center",
                        minWidth: `3rem`,
                        children: hasSectionAfter && /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
                            children: [
                                /*#__PURE__*/ jsxRuntime.jsx(Ornaments.VerticalDivider, {
                                    state: Ornaments.STATES.IS_DONE
                                }),
                                hasStepsBefore && /*#__PURE__*/ jsxRuntime.jsx(Ornaments.Number, {
                                    state: Ornaments.STATES.IS_ACTIVE,
                                    paddingTop: 3,
                                    children: nextSectionIndex + 1
                                })
                            ]
                        })
                    }),
                    /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Box, {
                        children: [
                            children,
                            cta && (cta.target ? /*#__PURE__*/ jsxRuntime.jsx(designSystem.LinkButton, {
                                tag: reactRouterDom.NavLink,
                                endIcon: /*#__PURE__*/ jsxRuntime.jsx(icons.ArrowRight, {}),
                                onClick: onCtaClick,
                                to: cta.target,
                                children: formatMessage(cta.title)
                            }) : /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                                endIcon: /*#__PURE__*/ jsxRuntime.jsx(icons.ArrowRight, {}),
                                onClick: onCtaClick,
                                children: formatMessage(cta.title)
                            }))
                        ]
                    })
                ]
            }),
            hasStepsBefore && hasSectionAfter && /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                paddingTop: 3,
                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                    marginRight: 8,
                    justifyContent: "center",
                    width: `3rem`,
                    children: /*#__PURE__*/ jsxRuntime.jsx(Ornaments.VerticalDivider, {
                        state: Ornaments.STATES.IS_DONE,
                        minHeight: `2.4rem`
                    })
                })
            })
        ]
    });
};
const GuidedTourContent = ({ id, defaultMessage })=>{
    const { formatMessage } = reactIntl.useIntl();
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
        direction: "column",
        alignItems: "stretch",
        gap: 4,
        paddingBottom: 6,
        children: formatMessage({
            id,
            defaultMessage
        }, {
            documentationLink: DocumentationLink,
            b: Bold,
            p: Paragraph,
            light: Light,
            ul: List,
            li: ListItem
        })
    });
};
const DocumentationLink = (children)=>/*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
        tag: "a",
        textColor: "primary600",
        target: "_blank",
        rel: "noopener noreferrer",
        href: "https://docs.strapi.io/developer-docs/latest/developer-resources/database-apis-reference/rest-api.html#api-parameters",
        children: children
    });
const Bold = (children)=>/*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
        fontWeight: "semiBold",
        children: children
    });
const Paragraph = (children)=>/*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
        children: children
    });
const Light = (children)=>/*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
        textColor: "neutral600",
        children: children
    });
const List = (children)=>/*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
        paddingLeft: 6,
        children: /*#__PURE__*/ jsxRuntime.jsx("ul", {
            children: children
        })
    });
const LiStyled = styledComponents.styled.li`
  list-style: disc;
  &::marker {
    color: ${({ theme })=>theme.colors.neutral800};
  }
`;
const ListItem = (children)=>/*#__PURE__*/ jsxRuntime.jsx(LiStyled, {
        children: children
    });

const MainNavWrapper = styledComponents.styled(designSystem.Flex)`
  border-right: 1px solid ${({ theme })=>theme.colors.neutral150};
`;
const MainNav = (props)=>/*#__PURE__*/ jsxRuntime.jsx(MainNavWrapper, {
        alignItems: "normal",
        tag: "nav",
        background: "neutral0",
        direction: "column",
        height: "100vh",
        position: "sticky",
        top: 0,
        zIndex: 2,
        width: 10,
        ...props
    });

const BrandIconWrapper = styledComponents.styled(designSystem.Flex)`
  svg,
  img {
    border-radius: ${({ theme })=>theme.borderRadius};
    object-fit: contain;
    height: 2.4rem;
    width: 2.4rem;
  }
`;
const NavBrand = ()=>{
    const { formatMessage } = reactIntl.useIntl();
    const { logos: { menu } } = Theme.useConfiguration('LeftMenu');
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
        padding: 3,
        children: /*#__PURE__*/ jsxRuntime.jsxs(BrandIconWrapper, {
            direction: "column",
            justifyContent: "center",
            width: "3.2rem",
            height: "3.2rem",
            children: [
                /*#__PURE__*/ jsxRuntime.jsx("img", {
                    src: menu.custom?.url || menu.default,
                    alt: formatMessage({
                        id: 'app.components.LeftMenu.logo.alt',
                        defaultMessage: 'Application logo'
                    }),
                    width: "100%",
                    height: "100%"
                }),
                /*#__PURE__*/ jsxRuntime.jsxs(designSystem.VisuallyHidden, {
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx("span", {
                            children: formatMessage({
                                id: 'app.components.LeftMenu.navbrand.title',
                                defaultMessage: 'Strapi Dashboard'
                            })
                        }),
                        /*#__PURE__*/ jsxRuntime.jsx("span", {
                            children: formatMessage({
                                id: 'app.components.LeftMenu.navbrand.workplace',
                                defaultMessage: 'Workplace'
                            })
                        })
                    ]
                })
            ]
        })
    });
};

/* -------------------------------------------------------------------------------------------------
 * Link
 * -----------------------------------------------------------------------------------------------*/ const MainNavLinkWrapper = styledComponents.styled(reactRouterDom.NavLink)`
  text-decoration: none;
  display: flex;
  border-radius: ${({ theme })=>theme.borderRadius};
  background: ${({ theme })=>theme.colors.neutral0};
  color: ${({ theme })=>theme.colors.neutral500};
  position: relative;
  width: fit-content;
  padding-block: 0.6rem;
  padding-inline: 0.6rem;

  &:hover {
    svg path {
      fill: ${({ theme })=>theme.colors.neutral600};
    }
    background: ${({ theme })=>theme.colors.neutral100};
  }

  &.active {
    svg path {
      fill: ${({ theme })=>theme.colors.primary600};
    }
    background: ${({ theme })=>theme.colors.primary100};
  }
`;
const LinkImpl = ({ children, ...props })=>{
    return /*#__PURE__*/ jsxRuntime.jsx(MainNavLinkWrapper, {
        ...props,
        children: children
    });
};
/* -------------------------------------------------------------------------------------------------
 * Tooltip
 * -----------------------------------------------------------------------------------------------*/ const TooltipImpl = ({ children, label, position = 'right' })=>{
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Tooltip, {
        side: position,
        label: label,
        delayDuration: 0,
        children: /*#__PURE__*/ jsxRuntime.jsx("span", {
            children: children
        })
    });
};
/* -------------------------------------------------------------------------------------------------
 * Icon
 * -----------------------------------------------------------------------------------------------*/ const IconImpl = ({ label, children })=>{
    if (!children) {
        return null;
    }
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.AccessibleIcon, {
        label: label,
        children: children
    });
};
/* -------------------------------------------------------------------------------------------------
 * Badge
 * -----------------------------------------------------------------------------------------------*/ const CustomBadge = styledComponents.styled(designSystem.Badge)`
  /* override default badge styles to change the border radius of the Base element in the Design System */
  border-radius: ${({ theme })=>theme.spaces[10]};
  height: 2rem;
`;
const BadgeImpl = ({ children, label, ...props })=>{
    if (!children) {
        return null;
    }
    return /*#__PURE__*/ jsxRuntime.jsx(CustomBadge, {
        position: "absolute",
        top: "-0.8rem",
        left: "1.7rem",
        "aria-label": label,
        active: false,
        ...props,
        children: children
    });
};
/* -------------------------------------------------------------------------------------------------
 * EXPORTS
 * -----------------------------------------------------------------------------------------------*/ const NavLink = {
    Link: LinkImpl,
    Tooltip: TooltipImpl,
    Icon: IconImpl,
    Badge: BadgeImpl
};

const MenuTrigger = styledComponents.styled(designSystem.Menu.Trigger)`
  height: ${({ theme })=>theme.spaces[7]};
  width: ${({ theme })=>theme.spaces[7]};
  border: none;
  border-radius: 50%;
  // Removes inherited 16px padding
  padding: 0;
  // Prevent empty pixel from appearing below the main nav
  overflow: hidden;
`;
const MenuContent = styledComponents.styled(designSystem.Menu.Content)`
  left: ${({ theme })=>theme.spaces[3]};
`;
const MenuItem = styledComponents.styled(designSystem.Menu.Item)`
  & > span {
    width: 100%;
    display: flex;
    align-items: center;
    gap: ${({ theme })=>theme.spaces[3]};
    justify-content: space-between;
  }
`;
const MenuItemDanger = styledComponents.styled(MenuItem)`
  &:hover {
    ${({ theme })=>`
    background: ${theme.colors.danger100};
  `}
  }
`;
const NavUser = ({ children, initials, ...props })=>{
    const { formatMessage } = reactIntl.useIntl();
    const navigate = reactRouterDom.useNavigate();
    const logout = Theme.useAuth('Logout', (state)=>state.logout);
    const handleProfile = ()=>{
        navigate('/me');
    };
    const handleLogout = ()=>{
        logout();
        navigate('/auth/login');
    };
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
        justifyContent: "center",
        padding: 3,
        borderStyle: "solid",
        borderWidth: "1px 0 0 0",
        borderColor: "neutral150",
        ...props,
        children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Menu.Root, {
            children: [
                /*#__PURE__*/ jsxRuntime.jsxs(MenuTrigger, {
                    endIcon: null,
                    fullWidth: true,
                    justifyContent: "center",
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Avatar.Item, {
                            delayMs: 0,
                            fallback: initials
                        }),
                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.VisuallyHidden, {
                            tag: "span",
                            children: children
                        })
                    ]
                }),
                /*#__PURE__*/ jsxRuntime.jsxs(MenuContent, {
                    popoverPlacement: "top-center",
                    zIndex: 3,
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx(MenuItem, {
                            onSelect: handleProfile,
                            children: formatMessage({
                                id: 'global.profile',
                                defaultMessage: 'Profile'
                            })
                        }),
                        /*#__PURE__*/ jsxRuntime.jsxs(MenuItemDanger, {
                            onSelect: handleLogout,
                            color: "danger600",
                            children: [
                                formatMessage({
                                    id: 'app.components.LeftMenu.logout',
                                    defaultMessage: 'Logout'
                                }),
                                /*#__PURE__*/ jsxRuntime.jsx(icons.SignOut, {})
                            ]
                        })
                    ]
                })
            ]
        })
    });
};

const sortLinks = (links)=>{
    return links.sort((a, b)=>{
        // if no position is defined, we put the link in the position of the external plugins, before the plugins list
        const positionA = a.position ?? 6;
        const positionB = b.position ?? 6;
        if (positionA < positionB) {
            return -1;
        } else {
            return 1;
        }
    });
};
const NavLinkBadgeCounter = styledComponents.styled(NavLink.Badge)`
  span {
    color: ${({ theme })=>theme.colors.neutral0};
  }
`;
const NavLinkBadgeLock = styledComponents.styled(NavLink.Badge)`
  background-color: transparent;
`;
const NavListWrapper = styledComponents.styled(designSystem.Flex)`
  overflow-y: auto;
`;
const LeftMenu = ({ generalSectionLinks, pluginsSectionLinks })=>{
    const user = Theme.useAuth('AuthenticatedApp', (state)=>state.user);
    const { trackUsage } = Theme.useTracking();
    const { pathname } = reactRouterDom.useLocation();
    const userDisplayName = users.getDisplayName(user);
    const { formatMessage, locale } = reactIntl.useIntl();
    const formatter = designSystem.useCollator(locale, {
        sensitivity: 'base'
    });
    const initials = userDisplayName.split(' ').map((name)=>name.substring(0, 1)).join('').substring(0, 2);
    const handleClickOnLink = (destination)=>{
        trackUsage('willNavigate', {
            from: pathname,
            to: destination
        });
    };
    const listLinksAlphabeticallySorted = [
        ...pluginsSectionLinks,
        ...generalSectionLinks
    ].sort((a, b)=>formatter.compare(formatMessage(a.intlLabel), formatMessage(b.intlLabel)));
    const listLinks = sortLinks(listLinksAlphabeticallySorted);
    return /*#__PURE__*/ jsxRuntime.jsxs(MainNav, {
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(NavBrand, {}),
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Divider, {}),
            /*#__PURE__*/ jsxRuntime.jsx(NavListWrapper, {
                tag: "ul",
                gap: 3,
                direction: "column",
                flex: 1,
                paddingTop: 3,
                paddingBottom: 3,
                children: listLinks.length > 0 ? listLinks.map((link)=>{
                    const LinkIcon = link.icon;
                    const badgeContentLock = link?.licenseOnly ? /*#__PURE__*/ jsxRuntime.jsx(icons.Lightning, {
                        fill: "warning500"
                    }) : undefined;
                    const badgeContentNumeric = link.notificationsCount && link.notificationsCount > 0 ? link.notificationsCount.toString() : undefined;
                    const labelValue = formatMessage(link.intlLabel);
                    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                        tag: "li",
                        children: /*#__PURE__*/ jsxRuntime.jsx(NavLink.Tooltip, {
                            label: labelValue,
                            children: /*#__PURE__*/ jsxRuntime.jsxs(NavLink.Link, {
                                to: link.to,
                                onClick: ()=>handleClickOnLink(link.to),
                                "aria-label": labelValue,
                                children: [
                                    /*#__PURE__*/ jsxRuntime.jsx(NavLink.Icon, {
                                        label: labelValue,
                                        children: /*#__PURE__*/ jsxRuntime.jsx(LinkIcon, {
                                            width: "20",
                                            height: "20",
                                            fill: "neutral500"
                                        })
                                    }),
                                    badgeContentLock ? /*#__PURE__*/ jsxRuntime.jsx(NavLinkBadgeLock, {
                                        label: "locked",
                                        textColor: "neutral500",
                                        paddingLeft: 0,
                                        paddingRight: 0,
                                        children: badgeContentLock
                                    }) : badgeContentNumeric ? /*#__PURE__*/ jsxRuntime.jsx(NavLinkBadgeCounter, {
                                        label: badgeContentNumeric,
                                        backgroundColor: "primary600",
                                        width: "2.3rem",
                                        color: "neutral0",
                                        children: badgeContentNumeric
                                    }) : null
                                ]
                            })
                        })
                    }, link.to);
                }) : null
            }),
            /*#__PURE__*/ jsxRuntime.jsx(NavUser, {
                initials: initials,
                children: userDisplayName
            })
        ]
    });
};

/**
 * TODO: this isn't great, and we really should focus on fixing this.
 */ const PluginsInitializer = ({ children })=>{
    const appPlugins = Theme.useStrapiApp('PluginsInitializer', (state)=>state.plugins);
    const [{ plugins }, dispatch] = React__namespace.useReducer(reducer, initialState, ()=>init(appPlugins));
    const setPlugin = React__namespace.useRef((pluginId)=>{
        dispatch({
            type: 'SET_PLUGIN_READY',
            pluginId
        });
    });
    const hasApluginNotReady = Object.keys(plugins).some((plugin)=>plugins[plugin].isReady === false);
    /**
   *
   * I have spent some time trying to understand what is happening here, and wanted to
   * leave that knowledge for my future me:
   *
   * `initializer` is an undocumented property of the `registerPlugin` API. At the time
   * of writing it seems only to be used by the i18n plugin.
   *
   * How does it work?
   *
   * Every plugin that has an `initializer` component defined, receives the
   * `setPlugin` function as a component prop. In the case of i18n the plugin fetches locales
   * first and calls `setPlugin` with `pluginId` once they are loaded, which then triggers the
   * reducer of the admin app defined above.
   *
   * Once all plugins are set to `isReady: true` the app renders.
   *
   * This API is used to block rendering of the admin app. We should remove that in v5 completely
   * and make sure plugins can inject data into the global store before they are initialized, to avoid
   * having a new prop-callback based communication channel between plugins and the core admin app.
   *
   */ if (hasApluginNotReady) {
        const initializers = Object.keys(plugins).reduce((acc, current)=>{
            const InitializerComponent = plugins[current].initializer;
            if (InitializerComponent) {
                const key = plugins[current].pluginId;
                acc.push(/*#__PURE__*/ jsxRuntime.jsx(InitializerComponent, {
                    setPlugin: setPlugin.current
                }, key));
            }
            return acc;
        }, []);
        return /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
            children: [
                initializers,
                /*#__PURE__*/ jsxRuntime.jsx(Theme.Page.Loading, {})
            ]
        });
    }
    return children;
};
const initialState = {
    plugins: {}
};
const reducer = (state = initialState, action)=>immer.produce(state, (draftState)=>{
        switch(action.type){
            case 'SET_PLUGIN_READY':
                {
                    draftState.plugins[action.pluginId].isReady = true;
                    break;
                }
            default:
                return draftState;
        }
    });
/* -------------------------------------------------------------------------------------------------
 * Init state
 * -----------------------------------------------------------------------------------------------*/ const init = (plugins)=>{
    return {
        plugins
    };
};

const useMenu = (shouldUpdateStrapi)=>{
    const checkUserHasPermissions = Theme.useAuth('useMenu', (state)=>state.checkUserHasPermissions);
    const menu = Theme.useStrapiApp('useMenu', (state)=>state.menu);
    const permissions = Theme.useTypedSelector((state)=>state.admin_app.permissions);
    const [menuWithUserPermissions, setMenuWithUserPermissions] = React__namespace.useState({
        generalSectionLinks: [
            {
                icon: icons.House,
                intlLabel: {
                    id: 'global.home',
                    defaultMessage: 'Home'
                },
                to: '/',
                permissions: [],
                position: 0
            },
            {
                icon: icons.ShoppingCart,
                intlLabel: {
                    id: 'global.marketplace',
                    defaultMessage: 'Marketplace'
                },
                to: '/marketplace',
                permissions: permissions.marketplace?.main ?? [],
                position: 7
            },
            {
                icon: icons.Cog,
                intlLabel: {
                    id: 'global.settings',
                    defaultMessage: 'Settings'
                },
                to: '/settings',
                // Permissions of this link are retrieved in the init phase
                // using the settings menu
                permissions: [],
                notificationsCount: 0,
                position: 9
            }
        ],
        pluginsSectionLinks: [],
        isLoading: true
    });
    const generalSectionLinksRef = React__namespace.useRef(menuWithUserPermissions.generalSectionLinks);
    React__namespace.useEffect(()=>{
        async function applyMenuPermissions() {
            const authorizedPluginSectionLinks = await getPluginSectionLinks(menu, checkUserHasPermissions);
            const authorizedGeneralSectionLinks = await getGeneralLinks(generalSectionLinksRef.current, shouldUpdateStrapi, checkUserHasPermissions);
            setMenuWithUserPermissions((state)=>({
                    ...state,
                    generalSectionLinks: authorizedGeneralSectionLinks,
                    pluginsSectionLinks: authorizedPluginSectionLinks,
                    isLoading: false
                }));
        }
        applyMenuPermissions();
    }, [
        setMenuWithUserPermissions,
        generalSectionLinksRef,
        menu,
        permissions,
        shouldUpdateStrapi,
        checkUserHasPermissions
    ]);
    return menuWithUserPermissions;
};
/* -------------------------------------------------------------------------------------------------
 * Utils
 * -----------------------------------------------------------------------------------------------*/ const getGeneralLinks = async (generalSectionRawLinks, shouldUpdateStrapi = false, checkUserHasPermissions)=>{
    const generalSectionLinksPermissions = await Promise.all(generalSectionRawLinks.map(({ permissions })=>checkUserHasPermissions(permissions)));
    const authorizedGeneralSectionLinks = generalSectionRawLinks.filter((_, index)=>generalSectionLinksPermissions[index].length > 0);
    const settingsLinkIndex = authorizedGeneralSectionLinks.findIndex((obj)=>obj.to === '/settings');
    if (settingsLinkIndex === -1) {
        return [];
    }
    const authorizedGeneralLinksClone = cloneDeep(authorizedGeneralSectionLinks);
    authorizedGeneralLinksClone[settingsLinkIndex].notificationsCount = shouldUpdateStrapi ? 1 : 0;
    return authorizedGeneralLinksClone;
};
const getPluginSectionLinks = async (pluginsSectionRawLinks, checkUserHasPermissions)=>{
    const pluginSectionLinksPermissions = await Promise.all(pluginsSectionRawLinks.map(({ permissions })=>checkUserHasPermissions(permissions)));
    const authorizedPluginSectionLinks = pluginsSectionRawLinks.filter((_, index)=>pluginSectionLinksPermissions[index].length > 0);
    return authorizedPluginSectionLinks;
};

const { version: strapiVersion } = packageInfo;
const AdminLayout = ()=>{
    const setGuidedTourVisibility = Theme.useGuidedTour('AdminLayout', (state)=>state.setGuidedTourVisibility);
    const { formatMessage } = reactIntl.useIntl();
    const userInfo = Theme.useAuth('AuthenticatedApp', (state)=>state.user);
    const [userId, setUserId] = React__namespace.useState();
    const { showReleaseNotification } = Theme.useConfiguration('AuthenticatedApp');
    const { data: appInfo, isLoading: isLoadingAppInfo } = admin.useInformationQuery();
    const [tagName, setTagName] = React__namespace.useState(strapiVersion);
    React__namespace.useEffect(()=>{
        if (showReleaseNotification) {
            fetch('https://api.github.com/repos/strapi/strapi/releases/latest').then(async (res)=>{
                if (!res.ok) {
                    return;
                }
                const response = await res.json();
                if (!response.tag_name) {
                    throw new Error();
                }
                setTagName(response.tag_name);
            }).catch(()=>{
            /**
           * silence is golden & we'll use the strapiVersion as a fallback
           */ });
        }
    }, [
        showReleaseNotification
    ]);
    const userRoles = Theme.useAuth('AuthenticatedApp', (state)=>state.user?.roles);
    React__namespace.useEffect(()=>{
        if (userRoles) {
            const isUserSuperAdmin = userRoles.find(({ code })=>code === 'strapi-super-admin');
            if (isUserSuperAdmin && appInfo?.autoReload) {
                setGuidedTourVisibility(true);
            }
        }
    }, [
        userRoles,
        appInfo?.autoReload,
        setGuidedTourVisibility
    ]);
    React__namespace.useEffect(()=>{
        users.hashAdminUserEmail(userInfo).then((id)=>{
            if (id) {
                setUserId(id);
            }
        });
    }, [
        userInfo
    ]);
    const { trackUsage } = Theme.useTracking();
    const { isLoading: isLoadingMenu, generalSectionLinks, pluginsSectionLinks } = useMenu(checkLatestStrapiVersion(strapiVersion, tagName));
    /**
   * Make sure the event is only send once after accessing the admin panel
   * and not at runtime for example when regenerating the permissions with the ctb
   * or with i18n
   */ useOnce.useOnce(()=>{
        trackUsage('didAccessAuthenticatedAdministration');
    });
    // We don't need to wait for the release query to be fetched before rendering the plugins
    // however, we need the appInfos and the permissions
    if (isLoadingMenu || isLoadingAppInfo) {
        return /*#__PURE__*/ jsxRuntime.jsx(Theme.Page.Loading, {});
    }
    return /*#__PURE__*/ jsxRuntime.jsxs(Theme.AppInfoProvider, {
        ...appInfo,
        userId: userId,
        latestStrapiReleaseTag: tagName,
        shouldUpdateStrapi: checkLatestStrapiVersion(strapiVersion, tagName),
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(index.NpsSurvey, {}),
            /*#__PURE__*/ jsxRuntime.jsx(PluginsInitializer, {
                children: /*#__PURE__*/ jsxRuntime.jsx(reactDnd.DndProvider, {
                    backend: reactDndHtml5Backend.HTML5Backend,
                    children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Box, {
                        background: "neutral100",
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.SkipToContent, {
                                children: formatMessage({
                                    id: 'skipToContent',
                                    defaultMessage: 'Skip to content'
                                })
                            }),
                            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                alignItems: "flex-start",
                                children: [
                                    /*#__PURE__*/ jsxRuntime.jsx(LeftMenu, {
                                        generalSectionLinks: generalSectionLinks,
                                        pluginsSectionLinks: pluginsSectionLinks
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Box, {
                                        flex: 1,
                                        children: [
                                            /*#__PURE__*/ jsxRuntime.jsx(reactRouterDom.Outlet, {}),
                                            /*#__PURE__*/ jsxRuntime.jsx(GuidedTourModal, {})
                                        ]
                                    })
                                ]
                            })
                        ]
                    })
                })
            })
        ]
    });
};
const PrivateAdminLayout = ()=>{
    return /*#__PURE__*/ jsxRuntime.jsx(PrivateRoute.PrivateRoute, {
        children: /*#__PURE__*/ jsxRuntime.jsx(AdminLayout, {})
    });
};
const checkLatestStrapiVersion = (currentPackageVersion, latestPublishedVersion = '')=>{
    if (!valid(currentPackageVersion) || !valid(latestPublishedVersion)) {
        return false;
    }
    return lt(currentPackageVersion, latestPublishedVersion);
};

exports.AdminLayout = AdminLayout;
exports.PrivateAdminLayout = PrivateAdminLayout;
