// @strapi/admin/dist/admin/chunks/AuthResponse-Cu7HgPEK.mjs
import { jsx } from 'react/jsx-runtime';
import * as React from 'react';
import { useIntl } from 'react-intl';
import { useMatch, useNavigate } from 'react-router-dom';
import { d as useTypedDispatch, P as Page, o as login } from './Theme-C3oHrQcy.mjs';
import '@radix-ui/react-context';
import '@strapi/design-system';
import '@strapi/icons';
import '@strapi/icons/symbols';
import 'qs';
import './admin-BBS7sDVu.mjs';
import 'lodash/trimEnd';
import '@reduxjs/toolkit/query/react';
import 'lodash/fp/pipe';
import 'lodash/clone';
import 'lodash/toPath';
import 'lodash/isEqual';
import 'axios';
import 'immer';
import 'lodash/get';
import 'lodash/set';
import 'use-context-selector';
import 'lodash/defaultsDeep';
import 'react-redux';
import 'styled-components';
import '@reduxjs/toolkit';

const getCookieValue = (name)=>{
    let result = null;
    const cookieArray = document.cookie.split(';');
    cookieArray.forEach((cookie)=>{
        const [key, value] = cookie.split('=').map((item)=>item.trim());
        if (key.trim() === name) {
            result = decodeURIComponent(value);
        }
    });
    return result;
};
const deleteCookie = (name)=>{
    // Set the cookie to expire in the past
    document.cookie = name + '=; Path=/; Expires=Thu, 01 Jan 1970 00:00:01 GMT;';
};

const AuthResponse = ()=>{
    const match = useMatch('/auth/login/:authResponse');
    const { formatMessage } = useIntl();
    const navigate = useNavigate();
    const dispatch = useTypedDispatch();
    const redirectToOops = React.useCallback(()=>{
        navigate({
            pathname: '/auth/oops',
            search: `?info=${encodeURIComponent(formatMessage({
                id: 'Auth.form.button.login.providers.error',
                defaultMessage: 'We cannot connect you through the selected provider.'
            }))}`
        });
    }, [
        navigate,
        formatMessage
    ]);
    React.useEffect(()=>{
        if (match?.params.authResponse === 'error') {
            redirectToOops();
        }
        if (match?.params.authResponse === 'success') {
            const jwtToken = getCookieValue('jwtToken');
            if (jwtToken) {
                dispatch(login({
                    token: jwtToken
                }));
                deleteCookie('jwtToken');
                navigate('/auth/login');
            } else {
                redirectToOops();
            }
        }
    }, [
        dispatch,
        match,
        redirectToOops,
        navigate
    ]);
    return /*#__PURE__*/ jsx(Page.Loading, {});
};

export { AuthResponse };
