// @strapi/admin/dist/admin/chunks/EventsTable-B1ytvu7n.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var EditPage = require('./EditPage-BLnONQN3.js');
require('react');
require('@strapi/design-system');
require('react-intl');
require('react-router-dom');
require('./Theme-DDlJz91O.js');
require('@radix-ui/react-context');
require('@strapi/icons');
require('@strapi/icons/symbols');
require('qs');
require('./admin-udBiOv2o.js');
require('lodash/trimEnd');
require('@reduxjs/toolkit/query/react');
require('lodash/fp/pipe');
require('lodash/clone');
require('lodash/toPath');
require('lodash/isEqual');
require('axios');
require('immer');
require('lodash/get');
require('lodash/set');
require('use-context-selector');
require('lodash/defaultsDeep');
require('react-redux');
require('styled-components');
require('@reduxjs/toolkit');
require('./selectors-Bf-6Fmez.js');
require('yup');
require('./index-C8yOprVG.js');
require('react-dom/client');
require('invariant');
require('lodash/isFunction');
require('lodash/merge');
require('lodash/pick');
require('react-query');
require('./useEnterprise-IHd3AKd2.js');
require('lodash/camelCase');
require('fractional-indexing');
require('lodash/omit');
require('formik');
require('lodash/throttle');
require('./useWebhooks-CROc5Uno.js');

const eeTables = {
    'review-workflows': {
        'review-workflows': [
            'review-workflows.updateEntryStage'
        ]
    },
    releases: {
        releases: [
            'releases.publish'
        ]
    }
};
const getHeaders = (table)=>{
    switch(table){
        case 'review-workflows':
            return ()=>[
                    {
                        id: 'review-workflows.updateEntryStage',
                        defaultMessage: 'Stage Change'
                    }
                ];
        case 'releases':
            return ()=>[
                    {
                        id: 'releases.publish',
                        defaultMessage: 'Publish'
                    }
                ];
    }
};
const EventsTableEE = ()=>{
    return /*#__PURE__*/ jsxRuntime.jsxs(EditPage.Events.Root, {
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(EditPage.Events.Headers, {}),
            /*#__PURE__*/ jsxRuntime.jsx(EditPage.Events.Body, {}),
            Object.keys(eeTables).map((table)=>/*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx(EditPage.Events.Headers, {
                            getHeaders: getHeaders(table)
                        }),
                        /*#__PURE__*/ jsxRuntime.jsx(EditPage.Events.Body, {
                            providedEvents: eeTables[table]
                        })
                    ]
                }))
        ]
    });
};

exports.EventsTableEE = EventsTableEE;
