// @strapi/admin/dist/admin/chunks/Layout-B2bTkm3u.mjs
import { jsxs, jsx } from 'react/jsx-runtime';
import { useIntl } from 'react-intl';
import { useLocation, NavLink, useMatch, Navigate, Outlet } from 'react-router-dom';
import { p as SETTINGS_LINKS_CE, b as Layouts } from './index-GvH7eseL.mjs';
import { u as useAuth, p as useAppInfo, q as useStrapiApp, c as useTracking, P as Page } from './Theme-C3oHrQcy.mjs';
import * as React from 'react';
import sortBy from 'lodash/sortBy';
import { useSelector } from 'react-redux';
import { s as selectAdminPermissions } from './selectors-Dej_6h4x.mjs';
import { u as useEnterprise } from './useEnterprise-Dy8wMNV5.mjs';
import { SubNavLink, SubNav, SubNavHeader, SubNavSections, SubNavSection } from '@strapi/design-system';
import { Lightning } from '@strapi/icons';
import { styled } from 'styled-components';
import 'react-dom/client';
import 'invariant';
import 'lodash/isFunction';
import 'lodash/merge';
import 'lodash/pick';
import 'immer';
import 'react-query';
import '@strapi/icons/symbols';
import './admin-BBS7sDVu.mjs';
import 'lodash/trimEnd';
import '@reduxjs/toolkit/query/react';
import 'lodash/fp/pipe';
import 'qs';
import 'lodash/camelCase';
import 'yup';
import 'fractional-indexing';
import 'lodash/isEqual';
import 'lodash/omit';
import 'formik';
import '@reduxjs/toolkit';
import 'lodash/throttle';
import '@radix-ui/react-context';
import 'lodash/clone';
import 'lodash/toPath';
import 'axios';
import 'lodash/get';
import 'lodash/set';
import 'use-context-selector';
import 'lodash/defaultsDeep';

const formatLinks = (menu)=>menu.map((menuSection)=>{
        const formattedLinks = menuSection.links.map((link)=>({
                ...link,
                isDisplayed: false
            }));
        return {
            ...menuSection,
            links: formattedLinks
        };
    });
const useSettingsMenu = ()=>{
    const [{ isLoading, menu }, setData] = React.useState({
        isLoading: true,
        menu: []
    });
    const checkUserHasPermission = useAuth('useSettingsMenu', (state)=>state.checkUserHasPermissions);
    const shouldUpdateStrapi = useAppInfo('useSettingsMenu', (state)=>state.shouldUpdateStrapi);
    const settings = useStrapiApp('useSettingsMenu', (state)=>state.settings);
    const permissions = useSelector(selectAdminPermissions);
    /**
   * memoize the return value of this function to avoid re-computing it on every render
   * because it's used in an effect it ends up re-running recursively.
   */ const ceLinks = React.useMemo(()=>SETTINGS_LINKS_CE(), []);
    const { admin: adminLinks, global: globalLinks } = useEnterprise(ceLinks, async ()=>(await import('./index-GvH7eseL.mjs').then(function (n) { return n.O; })).SETTINGS_LINKS_EE(), {
        combine (ceLinks, eeLinks) {
            return {
                admin: [
                    ...eeLinks.admin,
                    ...ceLinks.admin
                ],
                global: [
                    ...ceLinks.global,
                    ...eeLinks.global
                ]
            };
        },
        defaultValue: {
            admin: [],
            global: []
        }
    });
    const addPermissions = React.useCallback((link)=>{
        if (!link.id) {
            throw new Error('The settings menu item must have an id attribute.');
        }
        return {
            ...link,
            permissions: permissions.settings?.[link.id]?.main ?? []
        };
    }, [
        permissions.settings
    ]);
    React.useEffect(()=>{
        const getData = async ()=>{
            const buildMenuPermissions = (sections)=>Promise.all(sections.reduce((acc, section, sectionIndex)=>{
                    const linksWithPermissions = section.links.map(async (link, linkIndex)=>({
                            hasPermission: (await checkUserHasPermission(link.permissions)).length > 0,
                            sectionIndex,
                            linkIndex
                        }));
                    return [
                        ...acc,
                        ...linksWithPermissions
                    ];
                }, []));
            const menuPermissions = await buildMenuPermissions(sections);
            setData((prev)=>{
                return {
                    ...prev,
                    isLoading: false,
                    menu: sections.map((section, sectionIndex)=>({
                            ...section,
                            links: section.links.map((link, linkIndex)=>{
                                const permission = menuPermissions.find((permission)=>permission.sectionIndex === sectionIndex && permission.linkIndex === linkIndex);
                                return {
                                    ...link,
                                    isDisplayed: Boolean(permission?.hasPermission)
                                };
                            })
                        }))
                };
            });
        };
        const { global, ...otherSections } = settings;
        const sections = formatLinks([
            {
                ...global,
                links: sortBy([
                    ...global.links,
                    ...globalLinks.map(addPermissions)
                ], (link)=>link.id).map((link)=>({
                        ...link,
                        hasNotification: link.id === '000-application-infos' && shouldUpdateStrapi
                    }))
            },
            {
                id: 'permissions',
                intlLabel: {
                    id: 'Settings.permissions',
                    defaultMessage: 'Administration Panel'
                },
                links: adminLinks.map(addPermissions)
            },
            ...Object.values(otherSections)
        ]);
        getData();
    }, [
        adminLinks,
        globalLinks,
        settings,
        shouldUpdateStrapi,
        addPermissions,
        checkUserHasPermission
    ]);
    return {
        isLoading,
        menu: menu.map((menuItem)=>({
                ...menuItem,
                links: menuItem.links.filter((link)=>link.isDisplayed)
            }))
    };
};

const CustomIcon = styled(Lightning)`
  right: 15px;
  position: absolute;
  bottom: 50%;
  transform: translateY(50%);

  path {
    fill: ${({ theme })=>theme.colors.warning500};
  }
`;
const Link = styled(SubNavLink)`
  &.active ${CustomIcon} {
    right: 13px;
  }
`;
const SettingsNav = ({ menu })=>{
    const { formatMessage } = useIntl();
    const { trackUsage } = useTracking();
    const { pathname } = useLocation();
    const filteredMenu = menu.filter((section)=>!section.links.every((link)=>link.isDisplayed === false));
    const sections = filteredMenu.map((section)=>{
        return {
            ...section,
            title: section.intlLabel,
            links: section.links.map((link)=>{
                return {
                    ...link,
                    title: link.intlLabel,
                    name: link.id
                };
            })
        };
    });
    const label = formatMessage({
        id: 'global.settings',
        defaultMessage: 'Settings'
    });
    const handleClickOnLink = (destination)=>()=>{
            trackUsage('willNavigate', {
                from: pathname,
                to: destination
            });
        };
    return /*#__PURE__*/ jsxs(SubNav, {
        "aria-label": label,
        children: [
            /*#__PURE__*/ jsx(SubNavHeader, {
                label: label
            }),
            /*#__PURE__*/ jsx(SubNavSections, {
                children: sections.map((section)=>/*#__PURE__*/ jsx(SubNavSection, {
                        label: formatMessage(section.intlLabel),
                        children: section.links.map((link)=>{
                            return /*#__PURE__*/ jsxs(Link, {
                                tag: NavLink,
                                withBullet: link.hasNotification,
                                to: link.to,
                                onClick: handleClickOnLink(link.to),
                                position: "relative",
                                children: [
                                    formatMessage(link.intlLabel),
                                    link?.licenseOnly && /*#__PURE__*/ jsx(CustomIcon, {
                                        width: "1.5rem",
                                        height: "1.5rem"
                                    })
                                ]
                            }, link.id);
                        })
                    }, section.id))
            })
        ]
    });
};

const Layout = ()=>{
    /**
   * This ensures we're capturing the settingId from the URL
   * but also lets any nesting after that pass.
   */ const match = useMatch('/settings/:settingId/*');
    const { formatMessage } = useIntl();
    const { isLoading, menu } = useSettingsMenu();
    // Since the useSettingsMenu hook can make API calls in order to check the links permissions
    // We need to add a loading state to prevent redirecting the user while permissions are being checked
    if (isLoading) {
        return /*#__PURE__*/ jsx(Page.Loading, {});
    }
    if (!match?.params.settingId) {
        return /*#__PURE__*/ jsx(Navigate, {
            to: "application-infos"
        });
    }
    return /*#__PURE__*/ jsxs(Layouts.Root, {
        sideNav: /*#__PURE__*/ jsx(SettingsNav, {
            menu: menu
        }),
        children: [
            /*#__PURE__*/ jsx(Page.Title, {
                children: formatMessage({
                    id: 'global.settings',
                    defaultMessage: 'Settings'
                })
            }),
            /*#__PURE__*/ jsx(Outlet, {})
        ]
    });
};

export { Layout };
