// @strapi/admin/dist/admin/chunks/ListView-BBs2mFq6.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var symbols = require('@strapi/icons/symbols');
var qs = require('qs');
var reactIntl = require('react-intl');
var reactRouterDom = require('react-router-dom');
var Theme = require('./Theme-DDlJz91O.js');
var index = require('./index-C8yOprVG.js');
var useOnce = require('./useOnce-CCEXokcy.js');
var apiTokens = require('./apiTokens-C7CY0RRl.js');
var constants = require('./constants-tbFuOuW4.js');
var Table = require('./Table-Ce395NnP.js');
require('@radix-ui/react-context');
require('./admin-udBiOv2o.js');
require('lodash/trimEnd');
require('@reduxjs/toolkit/query/react');
require('lodash/fp/pipe');
require('lodash/clone');
require('lodash/toPath');
require('lodash/isEqual');
require('axios');
require('immer');
require('lodash/get');
require('lodash/set');
require('use-context-selector');
require('lodash/defaultsDeep');
require('react-redux');
require('styled-components');
require('@reduxjs/toolkit');
require('react-dom/client');
require('invariant');
require('lodash/isFunction');
require('lodash/merge');
require('lodash/pick');
require('react-query');
require('./useEnterprise-IHd3AKd2.js');
require('lodash/camelCase');
require('yup');
require('fractional-indexing');
require('lodash/omit');
require('formik');
require('lodash/throttle');
require('./RelativeTime-ChFKZs4z.js');
require('date-fns');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);
var qs__namespace = /*#__PURE__*/_interopNamespaceDefault(qs);

const TABLE_HEADERS = [
    {
        name: 'name',
        label: {
            id: 'Settings.apiTokens.ListView.headers.name',
            defaultMessage: 'Name'
        },
        sortable: true
    },
    {
        name: 'description',
        label: {
            id: 'Settings.apiTokens.ListView.headers.description',
            defaultMessage: 'Description'
        },
        sortable: false
    },
    {
        name: 'createdAt',
        label: {
            id: 'Settings.apiTokens.ListView.headers.createdAt',
            defaultMessage: 'Created at'
        },
        sortable: false
    },
    {
        name: 'lastUsedAt',
        label: {
            id: 'Settings.apiTokens.ListView.headers.lastUsedAt',
            defaultMessage: 'Last used'
        },
        sortable: false
    }
];
const ListView = ()=>{
    const { formatMessage } = reactIntl.useIntl();
    const { toggleNotification } = Theme.useNotification();
    const permissions = Theme.useTypedSelector((state)=>state.admin_app.permissions.settings?.['api-tokens']);
    const { allowedActions: { canRead, canCreate, canDelete, canUpdate } } = Theme.useRBAC(permissions);
    const navigate = reactRouterDom.useNavigate();
    const { trackUsage } = Theme.useTracking();
    const startSection = Theme.useGuidedTour('ListView', (state)=>state.startSection);
    const { _unstableFormatAPIError: formatAPIError } = Theme.useAPIErrorHandler();
    React__namespace.useEffect(()=>{
        startSection('apiTokens');
    }, [
        startSection
    ]);
    React__namespace.useEffect(()=>{
        navigate({
            search: qs__namespace.stringify({
                sort: 'name:ASC'
            }, {
                encode: false
            })
        });
    }, [
        navigate
    ]);
    const headers = TABLE_HEADERS.map((header)=>({
            ...header,
            label: formatMessage(header.label)
        }));
    useOnce.useOnce(()=>{
        trackUsage('willAccessTokenList', {
            tokenType: constants.API_TOKEN_TYPE
        });
    });
    const { data: apiTokens$1 = [], isLoading, error } = apiTokens.useGetAPITokensQuery();
    React__namespace.useEffect(()=>{
        if (error) {
            toggleNotification({
                type: 'danger',
                message: formatAPIError(error)
            });
        }
    }, [
        error,
        formatAPIError,
        toggleNotification
    ]);
    React__namespace.useEffect(()=>{
        trackUsage('didAccessTokenList', {
            number: apiTokens$1.length,
            tokenType: constants.API_TOKEN_TYPE
        });
    }, [
        apiTokens$1,
        trackUsage
    ]);
    const [deleteToken] = apiTokens.useDeleteAPITokenMutation();
    const handleDelete = async (id)=>{
        try {
            const res = await deleteToken(id);
            if ('error' in res) {
                toggleNotification({
                    type: 'danger',
                    message: formatAPIError(res.error)
                });
                return;
            }
            trackUsage('didDeleteToken');
        } catch  {
            toggleNotification({
                type: 'danger',
                message: formatMessage({
                    id: 'notification.error',
                    defaultMessage: 'Something went wrong'
                })
            });
        }
    };
    return /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(Theme.Page.Title, {
                children: formatMessage({
                    id: 'Settings.PageTitle',
                    defaultMessage: 'Settings - {name}'
                }, {
                    name: 'API Tokens'
                })
            }),
            /*#__PURE__*/ jsxRuntime.jsx(index.Layouts.Header, {
                title: formatMessage({
                    id: 'Settings.apiTokens.title',
                    defaultMessage: 'API Tokens'
                }),
                subtitle: formatMessage({
                    id: 'Settings.apiTokens.description',
                    defaultMessage: 'List of generated tokens to consume the API'
                }),
                primaryAction: canCreate && /*#__PURE__*/ jsxRuntime.jsx(designSystem.LinkButton, {
                    tag: reactRouterDom.Link,
                    "data-testid": "create-api-token-button",
                    startIcon: /*#__PURE__*/ jsxRuntime.jsx(icons.Plus, {}),
                    size: "S",
                    onClick: ()=>trackUsage('willAddTokenFromList', {
                            tokenType: constants.API_TOKEN_TYPE
                        }),
                    to: "/settings/api-tokens/create",
                    children: formatMessage({
                        id: 'Settings.apiTokens.create',
                        defaultMessage: 'Create new API Token'
                    })
                })
            }),
            !canRead ? /*#__PURE__*/ jsxRuntime.jsx(Theme.Page.NoPermissions, {}) : /*#__PURE__*/ jsxRuntime.jsx(Theme.Page.Main, {
                "aria-busy": isLoading,
                children: /*#__PURE__*/ jsxRuntime.jsxs(index.Layouts.Content, {
                    children: [
                        apiTokens$1.length > 0 && /*#__PURE__*/ jsxRuntime.jsx(Table.Table, {
                            permissions: {
                                canRead,
                                canDelete,
                                canUpdate
                            },
                            headers: headers,
                            isLoading: isLoading,
                            onConfirmDelete: handleDelete,
                            tokens: apiTokens$1,
                            tokenType: constants.API_TOKEN_TYPE
                        }),
                        canCreate && apiTokens$1.length === 0 ? /*#__PURE__*/ jsxRuntime.jsx(designSystem.EmptyStateLayout, {
                            icon: /*#__PURE__*/ jsxRuntime.jsx(symbols.EmptyDocuments, {
                                width: "16rem"
                            }),
                            content: formatMessage({
                                id: 'Settings.apiTokens.addFirstToken',
                                defaultMessage: 'Add your first API Token'
                            }),
                            action: /*#__PURE__*/ jsxRuntime.jsx(designSystem.LinkButton, {
                                tag: reactRouterDom.Link,
                                variant: "secondary",
                                startIcon: /*#__PURE__*/ jsxRuntime.jsx(icons.Plus, {}),
                                to: "/settings/api-tokens/create",
                                children: formatMessage({
                                    id: 'Settings.apiTokens.addNewToken',
                                    defaultMessage: 'Add new API Token'
                                })
                            })
                        }) : null,
                        !canCreate && apiTokens$1.length === 0 ? /*#__PURE__*/ jsxRuntime.jsx(designSystem.EmptyStateLayout, {
                            icon: /*#__PURE__*/ jsxRuntime.jsx(symbols.EmptyDocuments, {
                                width: "16rem"
                            }),
                            content: formatMessage({
                                id: 'Settings.apiTokens.emptyStateLayout',
                                defaultMessage: 'You don’t have any content yet...'
                            })
                        }) : null
                    ]
                })
            })
        ]
    });
};
const ProtectedListView = ()=>{
    const permissions = Theme.useTypedSelector((state)=>state.admin_app.permissions.settings?.['api-tokens'].main);
    return /*#__PURE__*/ jsxRuntime.jsx(Theme.Page.Protect, {
        permissions: permissions,
        children: /*#__PURE__*/ jsxRuntime.jsx(ListView, {})
    });
};

exports.ListView = ListView;
exports.ProtectedListView = ProtectedListView;
