// @strapi/admin/dist/admin/chunks/Login-hloLe5XK.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var designSystem = require('@strapi/design-system');
var reactIntl = require('react-intl');
var styledComponents = require('styled-components');
var index = require('./index-C8yOprVG.js');
var Theme = require('./Theme-DDlJz91O.js');
var SSOProviders = require('./SSOProviders-BC_OI8Id.js');
require('react-dom/client');
require('react');
require('invariant');
require('lodash/isFunction');
require('lodash/merge');
require('lodash/pick');
require('react-router-dom');
require('immer');
require('react-redux');
require('react-query');
require('@strapi/icons');
require('@strapi/icons/symbols');
require('./useEnterprise-IHd3AKd2.js');
require('./admin-udBiOv2o.js');
require('lodash/trimEnd');
require('@reduxjs/toolkit/query/react');
require('lodash/fp/pipe');
require('qs');
require('lodash/camelCase');
require('yup');
require('fractional-indexing');
require('lodash/isEqual');
require('lodash/omit');
require('formik');
require('@reduxjs/toolkit');
require('lodash/throttle');
require('@radix-ui/react-context');
require('lodash/clone');
require('lodash/toPath');
require('axios');
require('lodash/get');
require('lodash/set');
require('use-context-selector');
require('lodash/defaultsDeep');

const DividerFull = styledComponents.styled(designSystem.Divider)`
  flex: 1;
`;
const LoginEE = (loginProps)=>{
    const { formatMessage } = reactIntl.useIntl();
    const { isLoading, data: providers = [] } = Theme.useGetProvidersQuery(undefined, {
        skip: !window.strapi.features.isEnabled(window.strapi.features.SSO)
    });
    if (!window.strapi.features.isEnabled(window.strapi.features.SSO) || !isLoading && providers.length === 0) {
        return /*#__PURE__*/ jsxRuntime.jsx(index.Login, {
            ...loginProps
        });
    }
    return /*#__PURE__*/ jsxRuntime.jsx(index.Login, {
        ...loginProps,
        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
            paddingTop: 7,
            children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                direction: "column",
                alignItems: "stretch",
                gap: 7,
                children: [
                    /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsx(DividerFull, {}),
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                                paddingLeft: 3,
                                paddingRight: 3,
                                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                    variant: "sigma",
                                    textColor: "neutral600",
                                    children: formatMessage({
                                        id: 'Auth.login.sso.divider'
                                    })
                                })
                            }),
                            /*#__PURE__*/ jsxRuntime.jsx(DividerFull, {})
                        ]
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(SSOProviders.SSOProviders, {
                        providers: providers,
                        displayAllProviders: false
                    })
                ]
            })
        })
    });
};

exports.LoginEE = LoginEE;
