// @strapi/admin/dist/admin/chunks/Login-CqGaI_Ig.mjs
import { jsx, jsxs } from 'react/jsx-runtime';
import { Divider, Box, Flex, Typography } from '@strapi/design-system';
import { useIntl } from 'react-intl';
import { styled } from 'styled-components';
import { d as Login } from './index-GvH7eseL.mjs';
import { l as useGetProvidersQuery } from './Theme-C3oHrQcy.mjs';
import { S as SSOProviders } from './SSOProviders-Dw_eYbX-.mjs';
import 'react-dom/client';
import 'react';
import 'invariant';
import 'lodash/isFunction';
import 'lodash/merge';
import 'lodash/pick';
import 'react-router-dom';
import 'immer';
import 'react-redux';
import 'react-query';
import '@strapi/icons';
import '@strapi/icons/symbols';
import './useEnterprise-Dy8wMNV5.mjs';
import './admin-BBS7sDVu.mjs';
import 'lodash/trimEnd';
import '@reduxjs/toolkit/query/react';
import 'lodash/fp/pipe';
import 'qs';
import 'lodash/camelCase';
import 'yup';
import 'fractional-indexing';
import 'lodash/isEqual';
import 'lodash/omit';
import 'formik';
import '@reduxjs/toolkit';
import 'lodash/throttle';
import '@radix-ui/react-context';
import 'lodash/clone';
import 'lodash/toPath';
import 'axios';
import 'lodash/get';
import 'lodash/set';
import 'use-context-selector';
import 'lodash/defaultsDeep';

const DividerFull = styled(Divider)`
  flex: 1;
`;
const LoginEE = (loginProps)=>{
    const { formatMessage } = useIntl();
    const { isLoading, data: providers = [] } = useGetProvidersQuery(undefined, {
        skip: !window.strapi.features.isEnabled(window.strapi.features.SSO)
    });
    if (!window.strapi.features.isEnabled(window.strapi.features.SSO) || !isLoading && providers.length === 0) {
        return /*#__PURE__*/ jsx(Login, {
            ...loginProps
        });
    }
    return /*#__PURE__*/ jsx(Login, {
        ...loginProps,
        children: /*#__PURE__*/ jsx(Box, {
            paddingTop: 7,
            children: /*#__PURE__*/ jsxs(Flex, {
                direction: "column",
                alignItems: "stretch",
                gap: 7,
                children: [
                    /*#__PURE__*/ jsxs(Flex, {
                        children: [
                            /*#__PURE__*/ jsx(DividerFull, {}),
                            /*#__PURE__*/ jsx(Box, {
                                paddingLeft: 3,
                                paddingRight: 3,
                                children: /*#__PURE__*/ jsx(Typography, {
                                    variant: "sigma",
                                    textColor: "neutral600",
                                    children: formatMessage({
                                        id: 'Auth.login.sso.divider'
                                    })
                                })
                            }),
                            /*#__PURE__*/ jsx(DividerFull, {})
                        ]
                    }),
                    /*#__PURE__*/ jsx(SSOProviders, {
                        providers: providers,
                        displayAllProviders: false
                    })
                ]
            })
        })
    });
};

export { LoginEE };
