// @strapi/content-manager/dist/admin/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('@strapi/icons');
var index = require('./chunks/index-SQ88CePz.js');
require('prismjs');
require('react/jsx-runtime');
require('@strapi/admin/strapi-admin');
require('react');
require('@strapi/design-system');
require('lodash/fp/mapValues');
require('react-intl');
require('react-router-dom');
require('styled-components');
require('yup');
require('fractional-indexing');
require('lodash/fp/pipe');
require('qs');
require('date-fns');
require('@reduxjs/toolkit');



exports.DocumentRBAC = index.DocumentRBAC;
exports.DocumentStatus = index.DocumentStatus;
exports.buildValidParams = index.buildValidParams;
exports.default = index.index;
exports.unstable_useContentManagerContext = index.useContentManagerContext;
exports.unstable_useDocument = index.useDocument;
exports.unstable_useDocumentActions = index.useDocumentActions;
exports.unstable_useDocumentLayout = index.useDocumentLayout;
exports.useDocumentRBAC = index.useDocumentRBAC;
